### PR TITLE
add zokrates_js crate

### PR DESCRIPTION
**Pull request by dark64**
Todo
- [ ] Add proper CI steps for building and publishing zokrates_js (This might be moved to another PR)

### CODE REVIEW

File: zokrates_core/src/proof_system/mod.rs [hunk 1]
```diff
@@ -21,6 +22,18 @@ impl SetupKeypair {
     }
 }
 
+impl Serialize for SetupKeypair {
```
Comment (Schaeff) [inline zokrates_core/src/proof_system/mod.rs]: Any reason `derive(Serialize)` cannot be used here?

File: zokrates_js/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+[package]
+name = "zokrates_js"
+version = "1.0.0"
```
Comment (Schaeff) [inline zokrates_js/Cargo.toml]: we started from `0.1.0` on other crates

File: zokrates_js/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,7 @@
+FROM blockchainit/rust-wasm-env:latest
```
Comment (Schaeff) [inline zokrates_js/Dockerfile]: What are we using Docker for for zokrates_js?

File: zokrates_js/LICENSE [hunk 4]
```diff
@@ -0,0 +1,165 @@
+                   GNU LESSER GENERAL PUBLIC LICENSE
```
Comment (Schaeff) [inline zokrates_js/LICENSE]: Can we rely on the repo's LICENSE ?

File: zokrates_js/README.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
+# zokrates-js
```
Comment (Schaeff) [inline zokrates_js/README.md]: This should be migrated to `zokrates_book` I reckon

File: zokrates_js/gulpfile.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
+/** general */
```
Comment (Schaeff) [inline zokrates_js/gulpfile.js]: Would be nice to have some comments on what this does, I assume fetching stdlib? What version etc

File: zokrates_js/js/utils.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
+export function getAbsolutePath(basePath, relativePath) {
```
Comment (Schaeff) [inline zokrates_js/js/utils.js]: can we have some tests for the absolute resolution?

File: zokrates_js/package.json [hunk 8]
```diff
@@ -0,0 +1,37 @@
+{
+  "name": "zokrates-js",
+  "main": "js/index.js",
+  "author": "Darko Macesic <darem966@gmail.com>, Edi Sinovcic <edi@blockchain-it.hr>",
+  "version": "1.0.17",
```
Comment (Schaeff) [inline zokrates_js/package.json]: do we want to keep versions synced?

File: zokrates_js/Cargo.toml (hunk 2 above)
Comment (dark64) [reply zokrates_js/Cargo.toml]: Changed to 0.1.0

File: zokrates_core/src/proof_system/mod.rs (hunk 1 above)
Comment (dark64) [reply zokrates_core/src/proof_system/mod.rs]: No, I removed the unnecessary serializer

File: zokrates_js/Dockerfile (hunk 3 above)
Comment (dark64) [reply zokrates_js/Dockerfile]: It's not needed anymore

File: zokrates_js/LICENSE (hunk 4 above)
Comment (dark64) [reply zokrates_js/LICENSE]: Yes

File: zokrates_js/gulpfile.js (hunk 6 above)
Comment (dark64) [reply zokrates_js/gulpfile.js]: Yes, it's used for fetching and packing standard library into a json file. I added a comment for more clarity.

File: zokrates_js/js/utils.js (hunk 7 above)
Comment (dark64) [reply zokrates_js/js/utils.js]: Added tests for absolute resoltion and stdlib resolving.

File: zokrates_js/package.json (hunk 8 above)
Comment (dark64) [reply zokrates_js/package.json]: I think it's not needed, but we can discuss this further.

**Comment by Schaeff** [comment]
I think pulling wasm-friendly will fix the build

File: zokrates_js/tests/stdlib.test.js [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+const stdlib = require('../stdlib.json');
+
+describe('stdlib', function() {
+
+    it('should resolve module from stdlib (1)', function() {
+        let basePath = 'hashes/sha256/512bitPacked.zok';
+        let relativePath = '../../utils/pack/pack128';
+
+        let absolutePath = utils.appendExtension(utils.getAbsolutePath(basePath, relativePath), '.zok');
+        assert.notEqual(stdlib[absolutePath], undefined);
```
Comment (Schaeff) [inline zokrates_js/tests/stdlib.test.js]: Here we're testing that it resolves to something, could we test that it resolves to the right thing? `utils/pack/pack128` I assume?

Comment (Schaeff) [reply zokrates_js/tests/stdlib.test.js]: Just saw that's the approach you take for non-stdlib, why doesn't it work here?

File: zokrates_book/src/SUMMARY.md [hunk 10]
```diff
@@ -21,3 +21,4 @@
 	- [Testing](reference/testing.md)
 
 - [Tutorial: Proof of preimage](./sha256example.md)
+- [ZoKrates JS](./zokrates_js.md)
```
Comment (Schaeff) [inline zokrates_book/src/SUMMARY.md]:
```suggestion
- [zokrates.js](./zokrates_js.md)
```

File: zokrates_book/src/zokrates_js.md [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+function importResolver(location, path) {
+  // implement your resolving logic here
+  return { 
+    source: "def main() -> (): return", 
+    location: path 
+  };
+}
+
+zokrates.initialize(importResolver).then(() => {
+    // we have to initialize wasm module before calling api functions
```
Comment (Schaeff) [inline zokrates_book/src/zokrates_js.md]:
```suggestion
    // we have to initialize the wasm module before calling api functions
```

File: zokrates_core/src/proof_system/mod.rs [hunk 12]
```diff
@@ -10,6 +10,7 @@ pub use self::bn128::PGHR13;
 
 use crate::ir;
 
+#[derive(Serialize)]
```
Comment (Schaeff) [inline zokrates_core/src/proof_system/mod.rs]: Can you explain why we `Serialize` but don't `Deserialize` ?

File: zokrates_js/tests/stdlib.test.js [hunk 13]
```diff
@@ -0,0 +1,30 @@
+const assert = require('assert');
```
Comment (Schaeff) [inline zokrates_js/tests/stdlib.test.js]: Can we make sure these tests run in CI?

**Comment by Schaeff** [comment]
@dark64 should we have changes to CI in this PR as well? Maybe merge #514 into this

File: zokrates_js/tests/stdlib.test.js (hunk 9 above)
Comment (edisinovcic) [reply zokrates_js/tests/stdlib.test.js]: @Schaeff how would you test it that it resolved to right thing?
Hardcoding content of pack128 in the test? (That doesn't make that much sense to me)
Also testing that path is okay is utils/pack/pack128 also doesn't make sense.

File: zokrates_js/package.json (hunk 8 above)
Comment (Schaeff) [reply zokrates_js/package.json]: added to todo

**Pull request merged**